Sprint 16

AlbertKhaydarov opened the pull request.

AlbertKhaydarov committed:
add Onboarding

AlbertKhaydarov committed:
add category selected

AlbertKhaydarov committed:
edit

AlbertKhaydarov committed:
edit

AlbertKhaydarov committed:
edit

AlbertKhaydarov committed:
edit

AlbertKhaydarov committed:
edit

AlbertKhaydarov committed:
edit

AlbertKhaydarov committed:
Merge pull request #3 from AlbertKhaydarov/sprint_16_MVVM_Practicum

Sprint 16 mvvm practicum add MVVM

AlbertKhaydarov committed:
check to review

AlbertKhaydarov commented:
MVVM сделал для TrackersCategories и NewTracker

helecta commented on Tracker/CreateNewTracker/NewTracker/Helper/NewTrackerObservable.swift:
```diff
@@ -0,0 +1,26 @@
+        self.value = value
+    }
+    
+    private var listener: ((T?) -> Void)?
```
Напомню, что код стоит организовывать следующим образом:

- Публичные переменные
- IBOutlet
- Приватные переменные
- Публичные методы(сначала `init`ы, `override`ы, и потом остальные)
- Приватные методы
- IBAction

Подробнее почитать можно тут: [Организация кода](https://github.com/tutu-ru/swift-style-guide?tab=readme-ov-file#Организация-кода)

AlbertKhaydarov replied:
выполнено

helecta commented on Tracker/CreateNewTracker/NewTracker/View/NewTrackerViewController.swift:
```diff
@@ -166,35 +168,44 @@ final class NewTrackerViewController: UIViewController {
         setupTableViewAndCollectionViewHeight()
         self.viewRouter = ViewRouter(viewController: self)
+        
+        viewModel = NewTrackerVCViewModel()
```
Обычно viewModel не создается внутри вьюконтроллера, а инициализируются ей при создании контроллера. Т.е в данном случае должно было бы быть так в `EventTypeViewController`:
```
@objc private func creatNewHabit(_ sender: UIButton) {
        let viewModel = NewTrackerVCViewModel()
        let newHabitViewController = NewTrackerViewController(viewModel: viewModel)
...
```
В будущем это облегчило бы тестирование, да и делает код более гибким и готовым к переиспользованию.

helecta replied:
Аналогично сделать стоит и для других вьюконтроллеров

AlbertKhaydarov replied:
Выполнено. Отдельное спасибо за этот комментарий. Наконец-то у меня встало на место понимание для чего это делается (понятно, что глупый был вопрос, но он мне только сейчас прояснился

helecta commented on Tracker/Helpers/Storage/UserDefaultsStorage.swift:
```diff
@@ -25,5 +26,14 @@ final class UserDefaultsStorage: UserDefaultsStorageProtocol {
+   
+      var isFirstLaunch: Bool {
+          get {
+              return storage.bool(forKey: StorageKeys.isFirstLaunchKey.rawValue)
```
Обрати внимание, что если нет значения для ключа, то bool(forKey:) вернет false. Т.е у тебя при первом запуске будет возвращено значение false и поэтому онбординг не будет показан.

AlbertKhaydarov replied:
выполнено

helecta commented on Tracker/Onboarding/OnboardingViewController.swift:
```diff
@@ -0,0 +1,96 @@
+
+final class OnboardingViewController: UIPageViewController {
+    
+    private lazy var pages: [UIViewController] = {
```
Надо запомнить правило: вью контроллеры не надо хранить ссылками в других вью-контроллерах, а создавать их по мере необходимости, то есть при навигации.
Например, когда ты настраиваешь UIPageViewController - используешь setViewControllers -в этот момент создать первый, а когда будешь навигировать каруселью ко второму (через методы делегата), создавать второй, и так далее.

AlbertKhaydarov replied:
Выполнено

helecta commented on Tracker/CreateNewTracker/NewTracker/View/NewTrackerViewController.swift:
```diff
@@ -246,33 +247,39 @@ final class NewTrackerViewController: UIViewController {
         guard let cell = collectionView.cellForItem(at: indexPath) else { return }
         indexPathForSelectedEmoji = indexPath
+        let selectedEmoji = emojiesCollection[indexPath.row]
+        viewModel?.selectedEmoji.value?.selectedEmoji = selectedEmoji
```
Обработку выбора эмоджи стоит сразу передавать во вьюМодель:
```
    private func handleEmojiSelection(at indexPath: IndexPath) {
        if let selectedIndexPath = indexPathForSelectedEmoji, let cell = collectionView.cellForItem(at: selectedIndexPath) {
            clearSelection(for: cell, at: selectedIndexPath)
        }
        guard let cell = collectionView.cellForItem(at: indexPath) else { return }
        viewModel?.handleEmojiSelection(at: indexPath)
// обрати внимание, метод переименновала в selectCell, т.к метод обязательно должен содержать глагол
        selectCell(cell, at: indexPath,
                     withColor: .ypLightGray,
                     cornerRadius: 16)
// лучше переименовать в updateCreateButtonState()
        createButtonIsEnabled()
    }
```

А свойства 
```
    private var indexPathForSelectedEmoji: IndexPath?
    private var indexPathForSelectedColor: IndexPath?
    
    private var nameTracker = false
    private var timeSheetIsEnable = false
    private var categoryIsEnable = false
    private var emojiSelectedIsEnable = false
    private var colorSelectedIsEnable = false
```
Перенести во вью модель

helecta replied:
nameTracker - лучше переименовать в `isTrackerNameEmpty`, что бы сразу было понятно что это Bool

helecta replied:
Если пользователь ввел название и заполнил все остальные поля, а потом возьмет и удалит название, то у тебя nameTracker не будет переведен в состояние false, кнопка Создать не изменит свое состояние и на нее можно будет нажать и будет создан трекер без имени.

AlbertKhaydarov replied:
выполнено

helecta commented on Tracker/CreateNewTracker/NewTracker/View/NewTrackerViewController.swift:
```diff
@@ -246,33 +247,39 @@ final class NewTrackerViewController: UIViewController {
         guard let cell = collectionView.cellForItem(at: indexPath) else { return }
         indexPathForSelectedColor = indexPath
+        let selectedColor = colorSelection[indexPath.row]
+        viewModel?.selectedColor.value?.selectedColors = selectedColor
```
Аналогично, обработку выбора цвета стоит сразу передавать во вьюМодель

AlbertKhaydarov replied:
выполнено

helecta commented on Tracker/CreateNewTracker/NewTracker/View/NewTrackerViewController.swift:
```diff
@@ -419,61 +426,83 @@ extension NewTrackerViewController: UITableViewDataSource {
+    }
+    
+    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        if let typeEvent = TypeEvents(rawValue: self.title ?? "") {
```
Удобнее было бы убрать это во вьюМодель, и один раз при инициализации определить typeEvent для вьюМодели и соответсвенно и для контроллера. В контроллере получать уже сразу всю нужную информацию, т.е модель сама отдает нужные данные в зависимости от того какой TypeEvents внутри нее. Инициализировать модель с нужным typeEvent можно тут:
```
    @objc private func creatNewHabit(_ sender: UIButton) {
        ...
    }
```

AlbertKhaydarov replied:
Выполнено

AlbertKhaydarov committed:
Перенес строку с private var до инициализатора по рекомендации в Tracker/CreateNewTracker/NewTracker/Helper/NewTrackerObservable.swift

Добавлена инициализация  в Tracker/CreateNewTracker/NewTracker/View/NewTrackerViewController.swift с  viewModel (контроллер инициализируется в EventTypeViewController), аналогично исправил инициализацию CategoriesTypeViewController

При первом запуске не показывается онбординг - отредактирован инициализатор для установки значения по умолчанию  в Tracker/Helpers/Storage/UserDefaultsStorage.swift

AlbertKhaydarov committed:
переделан Tracker/Onboarding/OnboardingViewController.swift - подключение вьюконтроллеров (убрал хранение ссылками )

AlbertKhaydarov committed:
Исправления по результатам ревью:
Tracker/CreateNewTracker/NewTracker/View/NewTrackerViewController.swift - обработку выбора эмоджи перенес во вьюмодель
 свойства:
   private var indexPathForSelectedEmoji: IndexPath?
    private var indexPathForSelectedColor: IndexPath?

    private var nameTracker = false
    private var timeSheetIsEnable = false
    private var categoryIsEnable = false
    private var emojiSelectedIsEnable = false
    private var colorSelectedIsEnable = false
перенес во вьюмодель
nameTracker - переименовал в isTrackerNameEmpty

Tracker/CreateNewTracker/NewTracker/View/NewTrackerViewController.swift - аналогично перенес обработку выбора цвета во вьюМодель

AlbertKhaydarov committed:
Tracker/CreateNewTracker/NewTracker/View/NewTrackerViewController.swift
Добавлена проверка на не пустое поле ввода наименования трекера, чтобы нельзя было ввести трекер без названия
Добавил проверку на срабатывали после первого введенного символа

Tracker/CreateNewTracker/NewTracker/View/NewTrackerViewController.swift
Все проверки if let typeEvent = TypeEvents(rawValue: self.title ?? "" перенес во вьюмодель, инициализирую модель с нужным typeEvent

NewCategoryTypeViewController
Экран создания категории  соответствует макету (изменил цвет фона поля ввода).
Добавил проверку на срабатывали после первого введенного символа

CategoriesTypeViewController
 Добавил сохранение отметки последней выбранной категории
